### PR TITLE
Universal/fragment fog

### PR DESCRIPTION
**Pull request by eh-unity**
### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
* Fix bad looking fog with large triangles by moving fog computation to fragment shader. 
* It's now based on viewspace instead of clipspace. Originally to make it more uniform and not require remapping, also view space pos was already available in some shaders.
But afterwards the clip space scaling was fixed to be similar on all platforms. 
* This just forces the fragment fog enabled which avoids variants, but it's easy enough to turn into option if necessary.

Tested some performance on my Mac by rendering 4k fullscreen foggy cube.
Virtually no difference in speed for Lit or Unlit shader.
For Lit couple microseconds diff in 1ms of work, vertex shader 10 less instructions, pixel shader 7 more instructions.
For Unlit (3x faster than Lit btw.) the diff was low tens of microseconds, but the fragment version was repeatedly faster. That's < 5% diff though and the noise seems to be tens of microseconds too, so within noise margin (fs fog was consistently a bit faster though).

---
### Testing status
Added a test case for fog.
Ran graphics tests locally.

https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Ffragment-fog

---
### Comments to reviewers
This PR contains the fix in this PR https://github.com/Unity-Technologies/Graphics/pull/400 for shadergraph.

### CODE REVIEW

**Review by Verasl — CHANGES_REQUESTED** [review]
Before
image
After
image

Looking good, but I must say we should do the shadergraph targets here(can notice the road has shadergraph and still doing per vertex fog) at the same time, for example the pass templates for Lit is here and needs converteing/updating https://github.com/Unity-Technologies/Graphics/blob/master/com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/PBRForwardPass.hlsl

**Comment by ellioman** [comment]
I'm adding GFX QA as a reviewer as we'll need some performance testing as well once things are approved by other reviewers.

**Comment by eh-unity** [comment]
> Looking good, but I must say we should do the shadergraph targets here(can notice the road has shadergraph and still doing per vertex fog) at the same time, for example the pass templates for Lit is here and needs converteing/updating https://github.com/Unity-Technologies/Graphics/blob/master/com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/PBRForwardPass.hlsl

There's a PR to fix  the shadergraph which has broken fog node.
https://github.com/Unity-Technologies/Graphics/pull/400

**Comment by eh-unity** [comment]
> Looking good, but I must say we should do the shadergraph targets here(can notice the road has shadergraph and still doing per vertex fog) at the same time, for example the pass templates for Lit is here and needs converteing/updating 
Fog should be fixed for shadergraph now. There's a bug in Fog node for Unlit, so that won't be fixed until PR https://github.com/Unity-Technologies/Graphics/pull/400 is merged.

**Comment by eh-unity** [comment]
Universal only change,
Timeout and problems with multicamera test output
https://yamato-artifactviewer.prd.cds.internal.unity3d.com/eb045f1c-e75e-4dfc-ae2b-675cbceccade%2FTest_report%2FUtrTestReport1%2FTestProjects%2FUniversalGraphicsTest%2Ftest-results/TestReportV1.html

**Comment by eh-unity** [comment]
Universal CI seems to be green, except a time out in win player build.

File: com.unity.render-pipelines.universal/ShaderLibrary/Core.hlsl
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/Core.hlsl:42]: This path is dead code isn't it? OpenGL doesn't support reversed Z.

File: com.unity.render-pipelines.universal/Shaders/BakedLit.shader
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/BakedLit.shader:73]: If _FOG_FRAGMENT is already defined in shader library why do we need to redefine here?

File: com.unity.render-pipelines.universal/Shaders/Unlit.shader
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/Unlit.shader:70]: In this case both fogCoord and uv are floats. Shouldn't we pack them in same interpolator? Or does the compiler do this automagically?

File: com.unity.render-pipelines.universal/Shaders/Unlit.shader
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/Unlit.shader:222]: Same interpolator question here.

**Review by phi-lira — APPROVED** [review]
LGTM. Minor comments.

File: com.unity.render-pipelines.universal/Shaders/Unlit.shader
Comment (nigeljw-unity) [reply com.unity.render-pipelines.universal/Shaders/Unlit.shader:70]: These should definitely get packed by the compiler for the load, but I suspect these would end up in separate barycentric ops, as they would be in separate slots as different varyings. There are other compiler optimizations where all bary ops can be optimized out of the frag shader but that one is a bit more convoluted. I don't remember there being a packing opt for the bary ops on adreno, but other platforms may do this implicitly for free as long as they are aligned properly as Erik has done. Since everything is a scalar architecture these days, not sure how much benefit we would see, but I still think it is an interesting idea and worth trying (as it may reduce reg pressure). Definitely worth a look at the generated IR, but this might be an opt you only see in the machine code.

**Comment by eh-unity** [comment]
Universal CI passes, except for the Build win dx11 standalone which seems to have a timeout problem.

File: com.unity.render-pipelines.universal/Shaders/Unlit.shader
Comment (nigeljw-unity) [reply com.unity.render-pipelines.universal/Shaders/Unlit.shader:70]: So apparently, on GCN, the reduction in export is where we will see the benefit if we are export bound (and not ALU bound). This will be less true for RDNA, as it is less likely to be export bound, since exports are done in 4 dwords in batches of 4 (color and position have their own separate paths so not included in this on RDNA). Though as I mentioned, we won't see a reduction in ALU ops for the bary interpolations, but exports if pressured will cost more than an ALU op. There may be some implicit packing of exports, but this is less clear (though I will have a solid response on this at some point). While in the unlit shader, this will matter less, but in general, I think Felipe's suggestion is a really good general practice where possible.

File: com.unity.render-pipelines.universal/ShaderLibrary/Core.hlsl
Comment (eh-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Core.hlsl:42]: Currently as far as I know. But was there so I left it there, as REVERSE_Z doesn't exactly mean OpenGL (even if that is the case in practice.)
Do you want remove it?

File: com.unity.render-pipelines.universal/Shaders/BakedLit.shader
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Shaders/BakedLit.shader:73]: We don't if it doesn't include the library (might not be the case, but perhaps it then just should).
Most likely it's just leftovers when I had them per shader. I'll double check it.

File: com.unity.render-pipelines.universal/Shaders/Unlit.shader
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Shaders/Unlit.shader:70]: FXC implies that those are merged, uv and fogCoord get the same register numbers. i.e. v1.xy = uv and v1.z = fogCoord.
It looks like at least on my machine those are appear to be packed on lower level too.

**Comment by eh-unity** [comment]
ABV passed, except for some build timeouts for some mobile platforms and dx11.
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Ffragment-fog/.yamato%252F_abv.yml%2523all_project_ci_trunk/4375478/job

I triggered another Universal run because of the define changes, just to make sure.

**Comment by ellioman** [comment]
Removing myself as a reviewer as we already have Felipe, Nigel, Andre and QA.

**Comment by alindmanUnity** [comment]
@eh-unity Can we actually roll the fix for shader graph into this PR? The pr you mentioned was from a customer, which will take much longer to get up to correct state for merging. If we take it into this PR it will be much easier for us to fix and test and merge

**Comment by eh-unity** [comment]
@alindmanUnity The PR https://github.com/Unity-Technologies/Graphics/pull/400 or its equivalent is rolled into this PR. 
It's in the ShaderGraphFunctions.hlsl.

**Review by Verasl — APPROVED** [review]
LGTM

**Review by nigeljw-unity — APPROVED** [review]
Looks good to me! +1/2

I still need more exposure to URP/SRP/trunk to feel comfortable giving larger changes a proper +1.

**Review by erikabar — APPROVED** [review]
tested frame time on Android devices, results are pretty similar to master branch: https://docs.google.com/spreadsheets/d/1BwQxOJXhzk4X6GeSvPqUlbuzKSFYtH_N_oIpxKsvtbg/edit#gid=1958641881&range=A1:D1

File: com.unity.shadergraph/Documentation~/Fog-Node.md
Comment (elizabeth-legros) [inline com.unity.shadergraph/Documentation~/Fog-Node.md:22]: Small wording change request - "Returns a value between 0 and 1, where 0 is no fog and 1 is full fog"

**Review by hdb-unity — APPROVED** [review]
Checked that changes are made to new test projects (_Terrain and _Foundation).
Also, discussed latest failed CI tests to make sure they will be addressed - please don't merge till they are fixed.

Approving that changes are made to correct test project(s).
Note: Did not test. Did not run tests locally.

**Review by erikabar — APPROVED** [review]
approving this again from gfx-qa-urp side

**Comment by eh-unity** [comment]
CustomRevision (known broken trunk), seems to be all green.

**Pull request merged**
